### PR TITLE
[Feat] Manual Pinned Programs

### LINKED ISSUE #405: [Feature]: Don't just completely remove the "shortcuts area" from the UI if no Shortcuts yet
### Is your feature request related to a problem?

When nothing is starred (⭐) yet (no Shortcuts yet in current Bottle), don't just completely remove the "Shortcuts area" from the UI.

This is an opportunity to help brand new users get started.

leave_space_here

### Describe the solution you'd like

Reserve the space for the Shortcuts that will probably soon appear anyway, and:
- if nothing has been installed yet, add helpful text for new users: "Shortcuts for newly installed programs will appear here (press the "Install" button to install new software)" (or something like that)
- otherwise just leave the area empty, or add instructions for adding new Shortcuts?

### Anything else?

Nope. Thanks for considering this.

### Issue Language

- [X] Yes my issue is written in English

### PR DESCRIPTION
**Pull request by PortableProgrammer**
Adds UI elements for manually pinning programs, useful for built-in Windows utilities, portable applications, applications that are not installed within the `Program Files` or `Program Files (x86)` directories, applications that don't create Start Menu shortcuts, etc.

Fixes #405 

In the main Bottle view:
* Toolbar button to pin a program
* Call to action and prominent `Pin Program` button when the pinned programs list is empty
Screenshot 2023-11-07 at 7 39 01 PM

<details><summary>Manual Pinning UI and Result Examples</summary>
<p>
Screenshot 2023-11-07 at 7 39 34 PM
Screenshot 2023-11-07 at 7 39 43 PM
</p>
</details> 

<details><summary>Pinned programs are intelligently added to the Installed Programs list for configuration parity.</summary>
<p>
Screenshot 2023-11-07 at 7 39 52 PM
</p>
</details> 

<details><summary>Attempts to pin the same program multiple times are detected and disallowed.</summary>
<p>
Screenshot 2023-11-07 at 7 40 41 PM
</p>
</details>

### CODE REVIEW

**Comment by PortableProgrammer** [comment]
Replaces #539. Not sure what happened to that rebase, but the number of changes is small enough that it was easier to create a new branch and PR.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:31]: We gotta start privatizing these variables that are internal

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:24]: This probably should be a `let`

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:53]: instead of `Spacer()`, you can use `.frame(maxWidth: .infinity, alignment: .leading)`

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:39]: We could probably get rid of all these wrapping `HStack`s and use `VStack(alignment: .leading)`

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:77]: you can probably get rid of pinPath state and do a `let pinPath = newPinURL.prettyPath()` right here. Might help in the long-run maintainability of this feature.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:113]: I believe you can get rid of this button, the alert will add its own system button

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: Good catch. Fixed in latest commit.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: Concur. Fixed in latest commit.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: Fixed in latest commit.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: Removed extraneous `HStack`s by utilizing this recommendation in the latest commit.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: Had to do a bit of refactoring, but you're right, `pinPath` wasn't truly necessary. With some more refactoring, was able to remove a few redundant state variables in the latest commit.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: You're right; this was an unfortunate copy/paste result. Good catch.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:38]: you have some helper tags here. You will see `.infinity` highlighted in Xcode. you should double click on it to select it and get rid of these helper tags

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:45]: Same here

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:113]: You should really just do this:
```
Text("pin.error.duplicate.\(url.lastPathComponent)")
```

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:108]: Don't need the `String(localized: ..`.  `.alert("pin.error.title",...` should do the trick here

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:80]: just a preference but I think you can make newPinURL optional and not display the path if nothing is selected. It's a bit confusing to see a path for something you haven't selected yet.

I'll leave it to your discretion as I'm not a UX designer:

You will have to fix a few places:
1. You use an alert with `presenting` to deal with the optional url. You can leave the actions blank for the default "ok"
2. remove the `onAppear`
3. Deal with optional in panel results. Putting `?` should do the trick
4. Coalesce (`??`) a default panel directory in case nothing is yet selected: (i.e. `panel.directoryURL = newPinURL ?? bottle.url.appending(path: "drive_c")`

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:58]: A bit heavy on the nesting. you can use `guard result == .OK, let url = panel.urls.first else { return }` to get rid of all this unnecessary nesting. 

Since this is purely a preference and there are no coding guidelines written i'll leave it to your discretion. Unless @IsaacMarovitz insists.

**Review by cuba — CHANGES_REQUESTED** [review]

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: Arg, rookie mistake. Fixed in latest commit.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: Arg, rookie mistake. Fixed in latest commit.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: Agreed, got carried away with explicit localization. Fixed in latest commit.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: I was waffling on this anyway, so I took this as the impetus to just do it.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: Thank you for reminding me about the `guard` pattern; utilized here and in the `Pin` button to handle the now-optional `newPinURL`.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: This was an interesting one; I'm not super strong in localization, so this forced me to really focus on it. Fixed in latest commit.

**Comment by IsaacMarovitz** [comment]
> I get it on main

Curious what that is, maybe something in the pinned program array that isn't getting displayed correctly? Worth checking what's in your Bottle plist.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (IsaacMarovitz) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:70]: This layout doesn't match what we have in `BottleCreationView`. I'm not to bothered about UI being perfect in the UI as I'll probably tweak things after merge anyway.

File: Whisky/Models/Bottle.swift
Comment (IsaacMarovitz) [inline Whisky/Models/Bottle.swift:121]: We probably want to check if this program even exists anymore, maybe we can make Program a fallible init or something. Currently moving/renaming a manually pinned file doesn't remove it from the menu which is kind of confusing for end user.

File: Whisky/Views/Bottle/BottleView.swift
Comment (IsaacMarovitz) [inline Whisky/Views/Bottle/BottleView.swift:145]: I'm not sure how I feel about the placement and icon of this button. `pin` to me doesn't really imply "Add a pin" and adding this would sort of clash with the bottle add button. We might need to rethink those buttons in general.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (IsaacMarovitz) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:59]: This can be achieved without using regex. URL APIs provide us `.pathExtension`.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (IsaacMarovitz) [inline Whisky/Views/Bottle/Pins/PinCreationView.swift:86]: `.contains` would be simpler

**Review by IsaacMarovitz — CHANGES_REQUESTED** [review]

File: Whisky/Views/Bottle/BottleView.swift
Comment (cuba) [reply Whisky/Views/Bottle/BottleView.swift:145]: I felt this way too a little bit. I would rather see a dummy icon with a plus with an add pin label below it rather than a toolbar button for this case. Then there is something to fill the list when there are not pins too.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (cuba) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift:59]: yah there is `url.deletingPathExtension()`

**Comment by cuba** [comment]
> > I get it on main
> 
> Curious what that is, maybe something in the pinned program array that isn't getting displayed correctly? Worth checking what's in your Bottle plist.

Yah I have a duplicate pin. I can fix it. I think I created the issue in one of my PRs

File: Whisky/Views/Bottle/BottleView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/BottleView.swift:145]: Something like this? It's using the toolbar text, but it can be whatever.

Screenshot 2023-11-10 at 9 53 49 PM

<details><summary>Code to produce the above</summary>
<p>

It uses just enough of the `PinsView` layout to make me want to try to abstract it, but I'm honestly not sure it's worth it.

```swift
LazyVGrid(columns: gridLayout, alignment: .center) {
    ForEach(bottle.settings.pins, id: \.url) { pin in
        PinsView(
            bottle: bottle, pin: pin, path: $path
        )
    }
    VStack {
        Group {
            Image(systemName: "app.dashed")
                  .resizable()
                  .overlay {
                      Image(systemName: "plus")
                          .resizable()
                          .frame(width: 16, height: 16)
                  }
        }
        .frame(width: 45, height: 45)
        Spacer()
        Text("pin.help")
            .multilineTextAlignment(.center)
            .lineLimit(2, reservesSpace: true)
    }
    .frame(width: 90, height: 90)
    .padding(10)
    .onTapGesture(count: 1) {
        showPinCreation.toggle()
    }
}
.padding()
```
</p>
</details>

File: Whisky/Models/Bottle.swift
Comment (PortableProgrammer) [reply Whisky/Models/Bottle.swift:121]: I put this in originally because I needed to configure an environment variable on one of my manually pinned programs, and couldn't since it wasn't in the `programs` array, so it wasn't as fleshed out as it could have been. The concept of configuring a manually pinned program could probably stand to be reviewed more in-depth anyway. If you want to continue with this type of implementation, I can add detection logic into the `Program` init.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift:70]: Fair enough. FWIW, I started with `BottleCreationView` as a base pattern and then diverged predominantly as a result of reviews on this PR. If you have a preferred pattern, I'm happy to revert back to it, or you can adjust as necessary after the fact.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift:59]: Yes, that is indeed much easier to maintain. Will be in the next commit.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift:86]: `existingProgram` is used to test for both existence as well as `.pinned` in this block. If we switch to `.contains` here, we will still have to perform a similar lookup later whenever `.contains` is `true`. Granted, this should be far less often than _every time_, so the perceived performance impact would be less.

File: Whisky/Views/Bottle/Pins/PinCreationView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/Pins/PinCreationView.swift]: Fixed in latest commit.

**Comment by PortableProgrammer** [comment]
Latest commit:
Screenshot 2023-11-11 at 9 14 39 AM
Screenshot 2023-11-11 at 9 16 42 AM

File: Whisky/Views/Bottle/BottleView.swift
Comment (PortableProgrammer) [reply Whisky/Views/Bottle/BottleView.swift]: This is in the latest commit.

**Comment by PortableProgrammer** [comment]
Screenshot 2023-11-11 at 10 38 51 AM

The extraneous ellipses can be resolved by adding `reservesSpace: true` to `.lineLimit` and removing the extra `"\n"` when displaying the pins in `PinsView`:
``` swift
Text(name)
    .multilineTextAlignment(.center)
    .lineLimit(2, reservesSpace: true)
```

The result is a little cleaner:
Screenshot 2023-11-11 at 10 44 41 AM

**Comment by IsaacMarovitz** [comment]
Seem's I've created some issues following the rebase

**Comment by IsaacMarovitz** [comment]
Screenshot 2023-11-12 at 17 30 42

Screenshot 2023-11-12 at 17 30 45

Pins are not being named correctly

**Comment by IsaacMarovitz** [comment]
Screenshot 2023-11-12 at 17 32 09

Seems easy to create a de-sync between the config file on disk and what the application *thinks* is available for pinning.

1. Pin a program
2. Unpin a program
3. Try to pin that program again

**Comment by PortableProgrammer** [comment]
I'll review momentarily.

**Comment by PortableProgrammer** [comment]
> Pins are not being named correctly

This seems to be the culprit:
`let newlyCreatedPin = Program(url: newPinURL, bottle: bottle)`

`name` is no longer a valid parameter and is always defaulted in `Program`:
`let name = url.lastPathComponent`

**Comment by PortableProgrammer** [comment]
This exposes a weakness in the way I was handling naming pins. Shouldn't be too difficult to fix.

**Comment by PortableProgrammer** [comment]
Pins not being named correctly is resolved in my local working copy. This also fixed most of the desync issues, but there's one lingering. Working on it now.

**Comment by PortableProgrammer** [comment]
There's still one desync issue after re-pinning the same program. I'm trying to trigger a reload after adding a pin with the below. This works the first time, but doesn't sync correctly on subsequent attempts until a full bottle reload is performed (e.g. change the active bottle in the UI).

``` swift
// Trigger a reload
bottle.settings.pins = bottle.settings.pins
bottle.updateInstalledPrograms()
```

**Comment by PortableProgrammer** [comment]
@IsaacMarovitz I believe I've finally found the desync that was causing me so much grief.

``` swift
// In BottleView.swift
ForEach(bottle.pinnedPrograms, id: \.program.url)
```
The collection created by this `ForEach` is keyed off the URL, which doesn't change when re-pinning the same app, so the view never updates; it retains its previous values. We don't see this behavior when pinning existing programs since the pin name is defaulted, and we don't see it when changing the name via the context menu because that's happening to the item in the collection itself.

I propose updating `bottle.pinnedPrograms` to add an `id` element to the tuple that is a unique combination of `pin.name` and `program.url`, in order to force an update when one or the other changes. I used a string combination for `id` but we could use a hash or anything else if there's a concern about performance impact here.

It's worth noting that SwiftLint complains about tuples with more than two elements; I've disabled the warning for now, but the ultimate fix would be to create a new struct or move `id` into `PinnedProgram`, since `pin.url` and `program.url` will always match.
 
``` swift
// In Bottle.swift
public var pinnedPrograms: [(pin: PinnedProgram, program: Program, id: String)] {
    return settings.pins.compactMap { pin in
        guard let program = programs.first(where: { $0.url == pin.url }) else { return nil }
        return (pin, program, "\(pin.name)//\(program.url)")
    }
}

// In BottleView.swift
LazyVGrid(columns: gridLayout, alignment: .center) {
    ForEach(bottle.pinnedPrograms, id: \.id) { pinnedProgram in
        PinsView(
            bottle: bottle, program: pinnedProgram.program, pin: pinnedProgram.pin, path: $path
        )
    }
...
```

**Comment by PortableProgrammer** [comment]
With the latest commit, refactored and simplified a bit. `NSOpenPanel` is now modal in the Pin views, added animation to the new `PinButtonView`.

**Review by IsaacMarovitz — APPROVED** [review]

**Pull request merged**
